pest/examples/parens.rs

```diff
@@ -0,0 +1,105 @@
+extern crate pest;
+
+use std::io::{self, Write};
+use std::rc::Rc;
+
+use pest::inputs::{Input, Position, StringInput};
+use pest::iterators::Pairs;
+use pest::{Error, Parser, ParserState, state};
+
+#[allow(non_camel_case_types)]
+#[allow(dead_code)]
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+enum Rule {
+    expr,
+    paren,
+    paren_end
+}
+
+struct ParenParser;
+
+impl Parser<Rule> for ParenParser {
+    fn parse<I: Input>(rule: Rule, input: Rc<I>) -> Result<Pairs<Rule, I>, Error<Rule, I>> {
+        fn expr<I: Input>(
```

Comment:
I totally agree, but I suppose it will be easier to generate like this from the derive. The point is that the user should be able to implement whatever methods he wants on his derived `struct` and defining them outside of this function will flood *some* kind of namespace whichever way you take it.